File: copper-rx2/src/androidTest/java/app/cash/copper/rx2/AsRowsTest.java
package app.cash.copper.rx2;

import android.database.Cursor;
import androidx.test.runner.AndroidJUnit4;
import app.cash.copper.Query;
import app.cash.copper.testing.Employee;
import app.cash.copper.testing.NullQuery;
import java.util.concurrent.atomic.AtomicInteger;
import kotlin.jvm.functions.Function1;
import org.junit.Test;
import org.junit.runner.RunWith;

import static app.cash.copper.testing.Employee.queryOf;
import static com.google.common.truth.Truth.assertThat;

@RunWith(AndroidJUnit4.class)
@SuppressWarnings("CheckResult")
public final class AsRowsTest {
  @Test public void asRowsEmpty() {
    RxContentResolver.asRows(queryOf(), Employee.MAPPER)
        .test()
        .assertNoValues()
        .assertComplete();
  }

  @Test public void asRows() {
    Query query = queryOf("alice", "Alice Allison", "bob", "Bob Bobberson");
    RxContentResolver.asRows(query, Employee.MAPPER)
        .test()
        .assertValueAt(0, new Employee("alice", "Alice Allison"))
        .assertValueAt(1, new Employee("bob", "Bob Bobberson"))
        .assertComplete();
  }

  @Test public void asRowsStopsWhenUnsubscribed() {
    final AtomicInteger count = new AtomicInteger();
    Function1<Cursor, Employee> mapper = c -> {
      count.incrementAndGet();
      return Employee.MAPPER.invoke(c);
    };

    Query query = queryOf("alice", "Alice Allison", "bob", "Bob Bobberson");
    RxContentResolver.asRows(query, mapper)
        .take(1)
        .test()
        .assertValue(new Employee("alice", "Alice Allison"))
        .assertComplete();
    assertThat(count.get()).isEqualTo(1);
  }

  @Test public void asRowsEmptyWhenNullCursor() {
    final AtomicInteger count = new AtomicInteger();
    Function1<Cursor, Employee> mapper = cursor -> {
      count.incrementAndGet();
      return Employee.MAPPER.invoke(cursor);
    };

    RxContentResolver.asRows(NullQuery.INSTANCE, mapper)
      .test()
      .assertNoValues()
      .assertComplete();

    assertThat(count.get()).isEqualTo(0);
  }
}

File: copper-flow/src/androidTest/java/app/cash/copper/flow/AsRowsTest.kt
package app.cash.copper.flow

import androidx.test.runner.AndroidJUnit4
import app.cash.copper.testing.Employee
import app.cash.copper.testing.Employee.Companion.queryOf
import app.cash.copper.testing.NullQuery
import app.cash.turbine.test
import com.google.common.truth.Truth.assertThat
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.take
import kotlinx.coroutines.runBlocking
import org.junit.Test
import org.junit.runner.RunWith
import kotlin.time.ExperimentalTime

@ExperimentalCoroutinesApi
@ExperimentalTime
@RunWith(AndroidJUnit4::class)
class AsRowsTest {
  @Test fun asRowsEmpty() = runBlocking {
    queryOf()
      .asRows(mapper = Employee.MAPPER)
      .test {
        expectComplete()
      }
  }

  @Test fun asRows() = runBlocking {
    queryOf("alice", "Alice Allison", "bob", "Bob Bobberson")
      .asRows(mapper = Employee.MAPPER)
      .test {
        assertThat(expectItem()).isEqualTo(Employee("alice", "Alice Allison"))
        assertThat(expectItem()).isEqualTo(Employee("bob", "Bob Bobberson"))
        expectComplete()
      }
  }

  @Test fun asRowsStopsWhenCanceled() = runBlocking {
    var count = 0
    queryOf("alice", "Alice Allison", "bob", "Bob Bobberson", "eve", "Eve Evenson")
      .asRows {
        count++
        Employee.MAPPER.invoke(it)
      }
      .take(1)
      .test {
        expectItem()
        expectComplete()
      }
    // This is 2 not 1 because of how the implementation works. It will always be N+1.
    assertThat(count).isEqualTo(2)
  }

  @Test fun asRowsEmptyWhenNullCursor() = runBlocking {
    var count = 0
    NullQuery
      .asRows {
        count++
        Employee.MAPPER.invoke(it)
      }
      .test {
        expectComplete()
      }
    assertThat(count).isEqualTo(0)
  }
}
